change.go

```diff
@@ -17,49 +25,660 @@ const (
+	b.WriteString("}")
+	return b.String()
+}
+
+type CreateChangeConsumerInput struct {
+	TableName string
+	StreamIDs []StreamID
+}
+
+type ChangeConsumerFactory interface {
+	CreateChangeConsumer(input CreateChangeConsumerInput) (ChangeConsumer, error)
 }
 
 type ChangeConsumer interface {
-	Consume(Change)
+	End()
+	Consume(change Change) error
+}
+
+func MakeChangeConsumerFactoryFromFunc(f ChangeConsumerFunc) ChangeConsumerFactory {
+	return &changeConsumerFuncInstanceFactory{f}
+}
+
+type changeConsumerFuncInstanceFactory struct {
+	f ChangeConsumerFunc
+}
+
+func (ccfif *changeConsumerFuncInstanceFactory) CreateChangeConsumer(input CreateChangeConsumerInput) (ChangeConsumer, error) {
+	return &changeConsumerFuncInstance{
+		tableName: input.TableName,
+		f:         ccfif.f,
+	}, nil
+}
+
+type changeConsumerFuncInstance struct {
+	tableName string
+	f         ChangeConsumerFunc
+}
+
+func (ccfi *changeConsumerFuncInstance) End() {} // TODO: Snapshot here?
+func (ccfi *changeConsumerFuncInstance) Consume(change Change) error {
+	return ccfi.f(ccfi.tableName, change)
+}
+
+type ChangeConsumerFunc func(tableName string, change Change) error
+
+type changeRowQuerier struct {
+	keyspaceName string
+	tableName    string
+	session      *gocql.Session
+
+	pkCondition string
+	bindArgs    []interface{}
+}
+
+func newChangeRowQuerier(session *gocql.Session, streams []StreamID, keyspaceName, tableName string) *changeRowQuerier {
+	var pkCondition string
+	if len(streams) == 1 {
+		pkCondition = "\"cdc$stream_id\" = ?"
+	} else {
+		pkCondition = "\"cdc$stream_id\" IN (?" + strings.Repeat(", ?", len(streams)-1) + ")"
+	}
+
+	bindArgs := make([]interface{}, len(streams)+2)
+	for i, stream := range streams {
+		bindArgs[i] = stream
+	}
+
+	return &changeRowQuerier{
+		keyspaceName: keyspaceName,
+		tableName:    tableName,
+		session:      session,
+
+		pkCondition: pkCondition,
+		bindArgs:    bindArgs,
+	}
+}
+
+func (crq *changeRowQuerier) queryRange(start gocql.UUID, end gocql.UUID) (*changeRowIterator, error) {
+	// We need metadata to check if there are any tuples
+	kmeta, err := crq.session.KeyspaceMetadata(crq.keyspaceName)
+	if err != nil {
+		return nil, err
+	}
+
+	tmeta, ok := kmeta.Tables[crq.tableName+cdcTableSuffix]
+	if !ok {
+		return nil, fmt.Errorf("no such table: %s.%s", crq.keyspaceName, crq.tableName)
+	}
+
+	var colNames []string
+	var tupleNames []string
+	for _, col := range tmeta.Columns {
+		if strings.HasPrefix(col.Type, "frozen<tuple<") || strings.HasPrefix(col.Type, "tuple<") {
+			tupleNames = append(tupleNames, col.Name)
+			colNames = append(colNames, fmt.Sprintf("writetime(%s)", escapeColumnNameIfNeeded(col.Name)))
+		}
+	}
+
+	if len(tupleNames) == 0 {
+		colNames = []string{"*"}
+	} else {
+		for name := range tmeta.Columns {
+			colNames = append(colNames, escapeColumnNameIfNeeded(name))
+		}
+	}
+
+	queryStr := fmt.Sprintf(
+		"SELECT %s FROM %s.%s%s WHERE %s AND \"cdc$time\" > ? AND \"cdc$time\" < ? BYPASS CACHE",
```

Comment:
We probably could, but consider a situation in which a tuple column is added to the table, or a table had tuples and some column of unspecified type was added/removed. `changeRowQuerier` exists for the duration of the `streamBatchReader`, so this change won't be noticed until we switch to the next generation. We need to check the table metadata and construct the query each time in order not to miss schema change.

If we don't care about changing schema, then probably it's OK to do it the way you suggest.

Reply:
Handling changing schema properly would be nice, but I don't think this is enough.

First, the metadata is taken from session, which means that it will probably be out of date (it is being updated asynchronously). The only way to get the current metadata for a given query is to take it from the query result, no?

Also, there seems to be a problem with `select *` prepared statements if schema changes occur, at least according to DataStax docs:
https://docs.datastax.com/en/developer/java-driver/3.3/manual/statements/prepared/#avoid-preparing-select-queries
https://issues.apache.org/jira/browse/CASSANDRA-10786
although I don't know if Gocql automatically prepares `select *` statements too.

Finally, if we want to handle schema changes, the library should push probably a schema change event to each consumer, and the placing of the event should be consistent with other events, i.e. all previous events (changes) use the old schema and all later events use the new schema. Or something like that.